nexus/db-queries/src/db/datastore/target_release.rs

```diff
@@ -0,0 +1,265 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! [`DataStore`] methods to get/set the current [`TargetRelease`].
+
+use super::DataStore;
+use crate::authz;
+use crate::context::OpContext;
+use crate::db::error::{public_error_from_diesel, ErrorHandler};
+use crate::db::model::{SemverVersion, TargetRelease, TargetReleaseSource};
+use crate::db::schema::target_release;
+use crate::db::schema::target_release::dsl;
+use async_bb8_diesel::AsyncRunQueryDsl as _;
+use diesel::insert_into;
+use diesel::prelude::*;
+use nexus_types::external_api::views;
+use omicron_common::api::external::{CreateResult, Error, LookupResult};
+
+impl DataStore {
+    /// Fetch the current target release, i.e., the row with the largest
+    /// generation number.
+    pub async fn target_release_get_current(
+        &self,
+        opctx: &OpContext,
+    ) -> LookupResult<TargetRelease> {
+        opctx
+            .authorize(authz::Action::Read, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        // Fetch the row in the `target_release` table with the largest
+        // generation number. The subquery accesses the same table, so we
+        // have to make an alias to not confuse diesel.
+        let target_release_2 =
+            diesel::alias!(target_release as target_release_2);
+        dsl::target_release
+            .filter(dsl::generation.nullable().eq_any(target_release_2.select(
+                diesel::dsl::max(target_release_2.field(dsl::generation)),
+            )))
+            .select(TargetRelease::as_select())
+            .first_async(&*conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Insert a new target release row and return it. It will only become
+    /// the current target release if its generation is larger than any
+    /// existing row.
+    pub async fn target_release_insert(
+        &self,
+        opctx: &OpContext,
+        target_release: TargetRelease,
+    ) -> CreateResult<TargetRelease> {
+        opctx
+            .authorize(authz::Action::Modify, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+        self.transaction_retry_wrapper("set_target_release")
+            .transaction(&conn, |conn| {
```

Comment:
Since this is a single statement, I think it's better here to skip the explicit transaction.

If you skip `self.transaction_retry_wrapper(...).transaction()` altogether, then we'll just send this one statement to the database.  That's still treated as a transaction.  It's pretty unlikely to conflict with anything, but if it does, the database can retry it [automatically](https://www.cockroachlabs.com/docs/v25.1/transactions#automatic-retries).  This is generally better than a client-side retry because the latency is much smaller and the window for another conflict is much shorter.

Reply:
Thanks, also done in e7181dbbe.